// @netlify/dev/dist/main.js
// src/main.ts
import { promises as fs2 } from "fs";
import path2 from "path";
import process2 from "process";
import { resolveConfig } from "@netlify/config";
import { ensureNetlifyIgnore, getAPIToken, LocalState } from "@netlify/dev-utils";
import { FunctionsHandler } from "@netlify/functions/dev";
import { HeadersHandler } from "@netlify/headers";
import { RedirectsHandler } from "@netlify/redirects";
import { StaticHandler } from "@netlify/static";

// src/lib/env.ts
var SUPPORTED_CONTEXTS = ["all", "production", "deploy-preview", "branch-deploy", "dev"];
var injectEnvVariables = async ({
  accountSlug,
  baseVariables = {},
  envAPI,
  netlifyAPI,
  siteID
}) => {
  const results = {};
  let variables = baseVariables;
  if (netlifyAPI && accountSlug) {
    variables = await getEnvelopeEnv({
      accountId: accountSlug,
      api: netlifyAPI,
      env: baseVariables,
      siteId: siteID
    });
  }
  for (const [key, variable] of Object.entries(variables)) {
    const existsInProcess = envAPI.has(key);
    const [usedSource, ...overriddenSources] = existsInProcess ? ["process", ...variable.sources] : variable.sources;
    const isInternal = variable.sources.includes("internal");
    const result = {
      isInternal,
      originalValue: envAPI.get(key),
      overriddenSources,
      usedSource
    };
    if (!existsInProcess || isInternal) {
      envAPI.set(key, variable.value);
    }
    results[key] = result;
  }
  return results;
};
var getValueForContext = (values, contextOrBranch) => {
  const isSupportedContext = SUPPORTED_CONTEXTS.includes(contextOrBranch);
  if (!isSupportedContext) {
    const valueMatchingAsBranch = values.find((val) => val.context_parameter === contextOrBranch);
    if (valueMatchingAsBranch != null) {
      return valueMatchingAsBranch;
    }
    const valueMatchingContext = values.find((val) => val.context === "all" || val.context === "branch-deploy");
    return valueMatchingContext ?? void 0;
  }
  const valueMatchingAsContext = values.find((val) => val.context === "all" || val.context === contextOrBranch);
  return valueMatchingAsContext ?? void 0;
};
var filterEnvBySource = (env, source) => Object.fromEntries(Object.entries(env).filter(([, variable]) => variable.sources[0] === source));
var fetchEnvelopeItems = async function({
  accountId,
  api,
  key,
  siteId
}) {
  if (accountId === void 0) {
    return [];
  }
  try {
    if (key) {
      const envelopeItem = await api.getEnvVar({ accountId, key, siteId });
      return [envelopeItem];
    }
    const envelopeItems = await api.getEnvVars({ accountId, siteId });
    return envelopeItems;
  } catch {
    return [];
  }
};
var formatEnvelopeData = ({
  context = "dev",
  envelopeItems = [],
  scope = "any",
  source
}) => envelopeItems.filter(({ values }) => Boolean(getValueForContext(values, context))).filter(({ scopes }) => scope === "any" ? true : scopes.includes(scope)).sort((left, right) => left.key.toLowerCase() < right.key.toLowerCase() ? -1 : 1).reduce((acc, cur) => {
  const val = getValueForContext(cur.values, context);
  if (val === void 0) {
    throw new TypeError(`failed to locate environment variable value for ${context} context`);
  }
  const { context: itemContext, context_parameter: branch, value } = val;
  return {
    ...acc,
    [cur.key]: {
      context: itemContext,
      branch,
      scopes: cur.scopes,
      sources: [source],
      value
    }
  };
}, {});
var getEnvelopeEnv = async ({
  accountId,
  api,
  context = "dev",
  env,
  key = "",
  raw = false,
  scope = "any",
  siteId
}) => {
  const [accountEnvelopeItems, siteEnvelopeItems] = await Promise.all([
    fetchEnvelopeItems({ api, accountId, key }),
    fetchEnvelopeItems({ api, accountId, key, siteId })
  ]);
  const accountEnv = formatEnvelopeData({ context, envelopeItems: accountEnvelopeItems, scope, source: "account" });
  const siteEnv = formatEnvelopeData({ context, envelopeItems: siteEnvelopeItems, scope, source: "ui" });
  if (raw) {
    const entries = Object.entries({ ...accountEnv, ...siteEnv });
    return entries.reduce(
      (obj, [envVarKey, metadata]) => ({
        ...obj,
        [envVarKey]: metadata.value
      }),
      {}
    );
  }
  const generalEnv = filterEnvBySource(env, "general");
  const internalEnv = filterEnvBySource(env, "internal");
  const addonsEnv = filterEnvBySource(env, "addons");
  const configFileEnv = filterEnvBySource(env, "configFile");
  const includeConfigEnvVars = /any|builds|post[-_]processing/.test(scope);
  return {
    ...generalEnv,
    ...accountEnv,
    ...includeConfigEnvVars ? addonsEnv : {},
    ...siteEnv,
    ...includeConfigEnvVars ? configFileEnv : {},
    ...internalEnv
  };
};

// src/lib/fs.ts
import { promises as fs } from "fs";
var isDirectory = async (path3) => {
  try {
    const stat = await fs.stat(path3);
    return stat.isDirectory();
  } catch {
  }
  return false;
};
var isFile = async (path3) => {
  try {
    const stat = await fs.stat(path3);
    return stat.isFile();
  } catch {
  }
  return false;
};

// src/lib/runtime.ts
import path from "path";
import process from "process";
import { BlobsServer } from "@netlify/blobs/server";
import { startRuntime } from "@netlify/runtime";
var restoreEnvironment = (snapshot) => {
  for (const key in snapshot) {
    if (snapshot[key] === void 0) {
      delete process.env[key];
    } else {
      process.env[key] = snapshot[key];
    }
  }
};
var getRuntime = async ({ blobs, deployID, projectRoot, siteID }) => {
  const blobsToken = Math.random().toString().slice(2);
  const blobsServer = blobs ? new BlobsServer({
    directory: path.join(projectRoot, ".netlify", "blobs-serve"),
    token: blobsToken
  }) : null;
  const blobsServerDetails = await blobsServer?.start();
  const envSnapshot = {};
  const env = {
    delete: (key) => {
      envSnapshot[key] = envSnapshot[key] || process.env[key];
      delete process.env[key];
    },
    get: (key) => process.env[key],
    has: (key) => Boolean(process.env[key]),
    set: (key, value) => {
      envSnapshot[key] = envSnapshot[key] || process.env[key];
      process.env[key] = value;
    },
    toObject: () => process.env
  };
  startRuntime({
    blobs: blobsServerDetails ? {
      edgeURL: `http://localhost:${blobsServerDetails.port}`,
      uncachedEdgeURL: `http://localhost:${blobsServerDetails.port}`,
      primaryRegion: "us-east-2",
      token: blobsToken
    } : void 0,
    cache: {
      getCacheAPIContext: () => null,
      purgeToken: ""
    },
    deployID,
    env,
    getRequestContext: () => null,
    siteID
  });
  return {
    env,
    stop: async () => {
      restoreEnvironment(envSnapshot);
      await blobsServer?.stop();
    }
  };
};

// src/main.ts
var notFoundHandler = async () => new Response("Not found", { status: 404 });
var NetlifyDev = class {
  #apiHost;
  #apiScheme;
  #apiToken;
  #config;
  #features;
  #logger;
  #projectRoot;
  #runtime;
  #siteID;
  constructor(options) {
    if (options.apiURL) {
      const apiURL = new URL(options.apiURL);
      this.#apiHost = apiURL.host;
      this.#apiScheme = apiURL.protocol.slice(0, -1);
    }
    this.#apiToken = options.apiToken;
    this.#features = {
      blobs: options.blobs?.enabled !== false,
      environmentVariables: options.environmentVariables?.enabled !== false,
      functions: options.functions?.enabled !== false,
      headers: options.headers?.enabled !== false,
      redirects: options.redirects?.enabled !== false,
      static: options.staticFiles?.enabled !== false
    };
    this.#logger = options.logger ?? globalThis.console;
    this.#projectRoot = options.projectRoot ?? process2.cwd();
  }
  async handleInEphemeralDirectory(request, destPath) {
    const userFunctionsPath = this.#config?.config.functionsDirectory ?? path2.join(this.#projectRoot, "netlify/functions");
    const userFunctionsPathExists = await isDirectory(userFunctionsPath);
    const functions = this.#features.functions ? new FunctionsHandler({
      config: this.#config,
      destPath,
      projectRoot: this.#projectRoot,
      settings: {},
      siteId: this.#siteID,
      timeouts: {},
      userFunctionsPath: userFunctionsPathExists ? userFunctionsPath : void 0
    }) : null;
    const headers = this.#features.headers ? new HeadersHandler({
      configPath: this.#config?.configPath,
      configHeaders: this.#config?.config.headers,
      projectDir: this.#projectRoot,
      publishDir: this.#config?.config.build.publish ?? void 0,
      logger: this.#logger
    }) : { handle: async (_request, response) => response };
    const redirects = this.#features.redirects ? new RedirectsHandler({
      configPath: this.#config?.configPath,
      configRedirects: this.#config?.config.redirects,
      jwtRoleClaim: "",
      jwtSecret: "",
      notFoundHandler,
      projectDir: this.#projectRoot
    }) : null;
    const staticFiles = this.#features.static ? new StaticHandler({
      directory: this.#config?.config.build.publish ?? this.#projectRoot
    }) : null;
    const functionMatch = await functions?.match(request);
    if (functionMatch) {
      if (functionMatch.preferStatic) {
        const staticMatch2 = await staticFiles?.match(request);
        if (staticMatch2) {
          const response = await staticMatch2.handle();
          return headers.handle(request, response);
        }
      }
      return functionMatch.handle(request);
    }
    const redirectMatch = await redirects?.match(request);
    if (redirectMatch) {
      const functionMatch2 = await functions?.match(new Request(redirectMatch.target));
      if (functionMatch2 && !functionMatch2.preferStatic) {
        return functionMatch2.handle(request);
      }
      const response = await redirects?.handle(request, redirectMatch, async (maybeStaticFile) => {
        const staticMatch2 = await staticFiles?.match(maybeStaticFile);
        if (!staticMatch2) return;
        return async () => {
          const response2 = await staticMatch2.handle();
          return headers.handle(new Request(redirectMatch.target), response2);
        };
      });
      if (response) {
        return response;
      }
    }
    const staticMatch = await staticFiles?.match(request);
    if (staticMatch) {
      const response = await staticMatch.handle();
      return headers.handle(request, response);
    }
  }
  async getConfig() {
    const configFilePath = path2.resolve(this.#projectRoot, "netlify.toml");
    const configFileExists = await isFile(configFilePath);
    const config = await resolveConfig({
      config: configFileExists ? configFilePath : void 0,
      context: "dev",
      cwd: process2.cwd(),
      host: this.#apiHost,
      offline: !this.#siteID,
      mode: "cli",
      repositoryRoot: this.#projectRoot,
      scheme: this.#apiScheme,
      siteId: this.#siteID,
      token: this.#apiToken
    });
    return config;
  }
  async handle(request) {
    const servePath = path2.join(this.#projectRoot, ".netlify", "functions-serve");
    await fs2.mkdir(servePath, { recursive: true });
    const destPath = await fs2.mkdtemp(path2.join(servePath, "_"));
    try {
      return await this.handleInEphemeralDirectory(request, destPath);
    } finally {
      try {
        await fs2.rm(destPath, { force: true, recursive: true });
      } catch {
      }
    }
  }
  get siteIsLinked() {
    return Boolean(this.#siteID);
  }
  async start() {
    await ensureNetlifyIgnore(this.#projectRoot, this.#logger);
    this.#apiToken = this.#apiToken ?? await getAPIToken();
    const state = new LocalState(this.#projectRoot);
    const siteID = state.get("siteId");
    this.#siteID = siteID;
    const config = await this.getConfig();
    this.#config = config;
    const runtime = await getRuntime({
      blobs: Boolean(this.#features.blobs),
      deployID: "0",
      projectRoot: this.#projectRoot,
      siteID: siteID ?? "0"
    });
    this.#runtime = runtime;
    if (this.#features.environmentVariables && siteID) {
      await injectEnvVariables({
        accountSlug: config?.siteInfo?.account_slug,
        baseVariables: config?.env || {},
        envAPI: runtime.env,
        netlifyAPI: config?.api,
        siteID
      });
    }
  }
  async stop() {
    await this.#runtime?.stop();
  }
};
export {
  NetlifyDev
};
