// @netlify/dev/dist/main.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/main.ts
var main_exports = {};
__export(main_exports, {
  NetlifyDev: () => NetlifyDev
});
module.exports = __toCommonJS(main_exports);
var import_node_fs2 = require("fs");
var import_node_path2 = __toESM(require("path"), 1);
var import_node_process2 = __toESM(require("process"), 1);
var import_config = require("@netlify/config");
var import_dev_utils = require("@netlify/dev-utils");
var import_dev = require("@netlify/functions/dev");
var import_headers = require("@netlify/headers");
var import_redirects = require("@netlify/redirects");
var import_static = require("@netlify/static");

// src/lib/env.ts
var SUPPORTED_CONTEXTS = ["all", "production", "deploy-preview", "branch-deploy", "dev"];
var injectEnvVariables = async ({
  accountSlug,
  baseVariables = {},
  envAPI,
  netlifyAPI,
  siteID
}) => {
  const results = {};
  let variables = baseVariables;
  if (netlifyAPI && accountSlug) {
    variables = await getEnvelopeEnv({
      accountId: accountSlug,
      api: netlifyAPI,
      env: baseVariables,
      siteId: siteID
    });
  }
  for (const [key, variable] of Object.entries(variables)) {
    const existsInProcess = envAPI.has(key);
    const [usedSource, ...overriddenSources] = existsInProcess ? ["process", ...variable.sources] : variable.sources;
    const isInternal = variable.sources.includes("internal");
    const result = {
      isInternal,
      originalValue: envAPI.get(key),
      overriddenSources,
      usedSource
    };
    if (!existsInProcess || isInternal) {
      envAPI.set(key, variable.value);
    }
    results[key] = result;
  }
  return results;
};
var getValueForContext = (values, contextOrBranch) => {
  const isSupportedContext = SUPPORTED_CONTEXTS.includes(contextOrBranch);
  if (!isSupportedContext) {
    const valueMatchingAsBranch = values.find((val) => val.context_parameter === contextOrBranch);
    if (valueMatchingAsBranch != null) {
      return valueMatchingAsBranch;
    }
    const valueMatchingContext = values.find((val) => val.context === "all" || val.context === "branch-deploy");
    return valueMatchingContext ?? void 0;
  }
  const valueMatchingAsContext = values.find((val) => val.context === "all" || val.context === contextOrBranch);
  return valueMatchingAsContext ?? void 0;
};
var filterEnvBySource = (env, source) => Object.fromEntries(Object.entries(env).filter(([, variable]) => variable.sources[0] === source));
var fetchEnvelopeItems = async function({
  accountId,
  api,
  key,
  siteId
}) {
  if (accountId === void 0) {
    return [];
  }
  try {
    if (key) {
      const envelopeItem = await api.getEnvVar({ accountId, key, siteId });
      return [envelopeItem];
    }
    const envelopeItems = await api.getEnvVars({ accountId, siteId });
    return envelopeItems;
  } catch {
    return [];
  }
};
var formatEnvelopeData = ({
  context = "dev",
  envelopeItems = [],
  scope = "any",
  source
}) => envelopeItems.filter(({ values }) => Boolean(getValueForContext(values, context))).filter(({ scopes }) => scope === "any" ? true : scopes.includes(scope)).sort((left, right) => left.key.toLowerCase() < right.key.toLowerCase() ? -1 : 1).reduce((acc, cur) => {
  const val = getValueForContext(cur.values, context);
  if (val === void 0) {
    throw new TypeError(`failed to locate environment variable value for ${context} context`);
  }
  const { context: itemContext, context_parameter: branch, value } = val;
  return {
    ...acc,
    [cur.key]: {
      context: itemContext,
      branch,
      scopes: cur.scopes,
      sources: [source],
      value
    }
  };
}, {});
var getEnvelopeEnv = async ({
  accountId,
  api,
  context = "dev",
  env,
  key = "",
  raw = false,
  scope = "any",
  siteId
}) => {
  const [accountEnvelopeItems, siteEnvelopeItems] = await Promise.all([
    fetchEnvelopeItems({ api, accountId, key }),
    fetchEnvelopeItems({ api, accountId, key, siteId })
  ]);
  const accountEnv = formatEnvelopeData({ context, envelopeItems: accountEnvelopeItems, scope, source: "account" });
  const siteEnv = formatEnvelopeData({ context, envelopeItems: siteEnvelopeItems, scope, source: "ui" });
  if (raw) {
    const entries = Object.entries({ ...accountEnv, ...siteEnv });
    return entries.reduce(
      (obj, [envVarKey, metadata]) => ({
        ...obj,
        [envVarKey]: metadata.value
      }),
      {}
    );
  }
  const generalEnv = filterEnvBySource(env, "general");
  const internalEnv = filterEnvBySource(env, "internal");
  const addonsEnv = filterEnvBySource(env, "addons");
  const configFileEnv = filterEnvBySource(env, "configFile");
  const includeConfigEnvVars = /any|builds|post[-_]processing/.test(scope);
  return {
    ...generalEnv,
    ...accountEnv,
    ...includeConfigEnvVars ? addonsEnv : {},
    ...siteEnv,
    ...includeConfigEnvVars ? configFileEnv : {},
    ...internalEnv
  };
};

// src/lib/fs.ts
var import_node_fs = require("fs");
var isDirectory = async (path3) => {
  try {
    const stat = await import_node_fs.promises.stat(path3);
    return stat.isDirectory();
  } catch {
  }
  return false;
};
var isFile = async (path3) => {
  try {
    const stat = await import_node_fs.promises.stat(path3);
    return stat.isFile();
  } catch {
  }
  return false;
};

// src/lib/runtime.ts
var import_node_path = __toESM(require("path"), 1);
var import_node_process = __toESM(require("process"), 1);
var import_server = require("@netlify/blobs/server");
var import_runtime = require("@netlify/runtime");
var restoreEnvironment = (snapshot) => {
  for (const key in snapshot) {
    if (snapshot[key] === void 0) {
      delete import_node_process.default.env[key];
    } else {
      import_node_process.default.env[key] = snapshot[key];
    }
  }
};
var getRuntime = async ({ blobs, deployID, projectRoot, siteID }) => {
  const blobsToken = Math.random().toString().slice(2);
  const blobsServer = blobs ? new import_server.BlobsServer({
    directory: import_node_path.default.join(projectRoot, ".netlify", "blobs-serve"),
    token: blobsToken
  }) : null;
  const blobsServerDetails = await blobsServer?.start();
  const envSnapshot = {};
  const env = {
    delete: (key) => {
      envSnapshot[key] = envSnapshot[key] || import_node_process.default.env[key];
      delete import_node_process.default.env[key];
    },
    get: (key) => import_node_process.default.env[key],
    has: (key) => Boolean(import_node_process.default.env[key]),
    set: (key, value) => {
      envSnapshot[key] = envSnapshot[key] || import_node_process.default.env[key];
      import_node_process.default.env[key] = value;
    },
    toObject: () => import_node_process.default.env
  };
  (0, import_runtime.startRuntime)({
    blobs: blobsServerDetails ? {
      edgeURL: `http://localhost:${blobsServerDetails.port}`,
      uncachedEdgeURL: `http://localhost:${blobsServerDetails.port}`,
      primaryRegion: "us-east-2",
      token: blobsToken
    } : void 0,
    cache: {
      getCacheAPIContext: () => null,
      purgeToken: ""
    },
    deployID,
    env,
    getRequestContext: () => null,
    siteID
  });
  return {
    env,
    stop: async () => {
      restoreEnvironment(envSnapshot);
      await blobsServer?.stop();
    }
  };
};

// src/main.ts
var notFoundHandler = async () => new Response("Not found", { status: 404 });
var NetlifyDev = class {
  #apiHost;
  #apiScheme;
  #apiToken;
  #config;
  #features;
  #logger;
  #projectRoot;
  #runtime;
  #siteID;
  constructor(options) {
    if (options.apiURL) {
      const apiURL = new URL(options.apiURL);
      this.#apiHost = apiURL.host;
      this.#apiScheme = apiURL.protocol.slice(0, -1);
    }
    this.#apiToken = options.apiToken;
    this.#features = {
      blobs: options.blobs?.enabled !== false,
      environmentVariables: options.environmentVariables?.enabled !== false,
      functions: options.functions?.enabled !== false,
      headers: options.headers?.enabled !== false,
      redirects: options.redirects?.enabled !== false,
      static: options.staticFiles?.enabled !== false
    };
    this.#logger = options.logger ?? globalThis.console;
    this.#projectRoot = options.projectRoot ?? import_node_process2.default.cwd();
  }
  async handleInEphemeralDirectory(request, destPath) {
    const userFunctionsPath = this.#config?.config.functionsDirectory ?? import_node_path2.default.join(this.#projectRoot, "netlify/functions");
    const userFunctionsPathExists = await isDirectory(userFunctionsPath);
    const functions = this.#features.functions ? new import_dev.FunctionsHandler({
      config: this.#config,
      destPath,
      projectRoot: this.#projectRoot,
      settings: {},
      siteId: this.#siteID,
      timeouts: {},
      userFunctionsPath: userFunctionsPathExists ? userFunctionsPath : void 0
    }) : null;
    const headers = this.#features.headers ? new import_headers.HeadersHandler({
      configPath: this.#config?.configPath,
      configHeaders: this.#config?.config.headers,
      projectDir: this.#projectRoot,
      publishDir: this.#config?.config.build.publish ?? void 0,
      logger: this.#logger
    }) : { handle: async (_request, response) => response };
    const redirects = this.#features.redirects ? new import_redirects.RedirectsHandler({
      configPath: this.#config?.configPath,
      configRedirects: this.#config?.config.redirects,
      jwtRoleClaim: "",
      jwtSecret: "",
      notFoundHandler,
      projectDir: this.#projectRoot
    }) : null;
    const staticFiles = this.#features.static ? new import_static.StaticHandler({
      directory: this.#config?.config.build.publish ?? this.#projectRoot
    }) : null;
    const functionMatch = await functions?.match(request);
    if (functionMatch) {
      if (functionMatch.preferStatic) {
        const staticMatch2 = await staticFiles?.match(request);
        if (staticMatch2) {
          const response = await staticMatch2.handle();
          return headers.handle(request, response);
        }
      }
      return functionMatch.handle(request);
    }
    const redirectMatch = await redirects?.match(request);
    if (redirectMatch) {
      const functionMatch2 = await functions?.match(new Request(redirectMatch.target));
      if (functionMatch2 && !functionMatch2.preferStatic) {
        return functionMatch2.handle(request);
      }
      const response = await redirects?.handle(request, redirectMatch, async (maybeStaticFile) => {
        const staticMatch2 = await staticFiles?.match(maybeStaticFile);
        if (!staticMatch2) return;
        return async () => {
          const response2 = await staticMatch2.handle();
          return headers.handle(new Request(redirectMatch.target), response2);
        };
      });
      if (response) {
        return response;
      }
    }
    const staticMatch = await staticFiles?.match(request);
    if (staticMatch) {
      const response = await staticMatch.handle();
      return headers.handle(request, response);
    }
  }
  async getConfig() {
    const configFilePath = import_node_path2.default.resolve(this.#projectRoot, "netlify.toml");
    const configFileExists = await isFile(configFilePath);
    const config = await (0, import_config.resolveConfig)({
      config: configFileExists ? configFilePath : void 0,
      context: "dev",
      cwd: import_node_process2.default.cwd(),
      host: this.#apiHost,
      offline: !this.#siteID,
      mode: "cli",
      repositoryRoot: this.#projectRoot,
      scheme: this.#apiScheme,
      siteId: this.#siteID,
      token: this.#apiToken
    });
    return config;
  }
  async handle(request) {
    const servePath = import_node_path2.default.join(this.#projectRoot, ".netlify", "functions-serve");
    await import_node_fs2.promises.mkdir(servePath, { recursive: true });
    const destPath = await import_node_fs2.promises.mkdtemp(import_node_path2.default.join(servePath, "_"));
    try {
      return await this.handleInEphemeralDirectory(request, destPath);
    } finally {
      try {
        await import_node_fs2.promises.rm(destPath, { force: true, recursive: true });
      } catch {
      }
    }
  }
  get siteIsLinked() {
    return Boolean(this.#siteID);
  }
  async start() {
    await (0, import_dev_utils.ensureNetlifyIgnore)(this.#projectRoot, this.#logger);
    this.#apiToken = this.#apiToken ?? await (0, import_dev_utils.getAPIToken)();
    const state = new import_dev_utils.LocalState(this.#projectRoot);
    const siteID = state.get("siteId");
    this.#siteID = siteID;
    const config = await this.getConfig();
    this.#config = config;
    const runtime = await getRuntime({
      blobs: Boolean(this.#features.blobs),
      deployID: "0",
      projectRoot: this.#projectRoot,
      siteID: siteID ?? "0"
    });
    this.#runtime = runtime;
    if (this.#features.environmentVariables && siteID) {
      await injectEnvVariables({
        accountSlug: config?.siteInfo?.account_slug,
        baseVariables: config?.env || {},
        envAPI: runtime.env,
        netlifyAPI: config?.api,
        siteID
      });
    }
  }
  async stop() {
    await this.#runtime?.stop();
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  NetlifyDev
});
