// @netlify/vite-plugin/dist/main.js
// src/main.ts
import process from "process";
import { NetlifyDev } from "@netlify/dev";

// src/lib/logger.ts
import chalk from "chalk";
var NETLIFY_CYAN = chalk.rgb(40, 180, 170);
var banner = NETLIFY_CYAN("\u2B25 Netlify \u2B25");
var logger = {
  error: (...data) => data.length === 0 ? console.error(...data) : console.error(banner, ...data),
  log: (...data) => data.length === 0 ? console.log(...data) : console.log(banner, ...data),
  warn: (...data) => data.length === 0 ? console.warn(...data) : console.warn(banner, ...data)
};

// src/lib/reqres.ts
import { Readable } from "stream";
var normalizeHeaders = (headers) => {
  const result = [];
  for (const [key, value] of Object.entries(headers)) {
    if (Array.isArray(value)) {
      result.push([key, value.join(",")]);
    } else if (typeof value === "string") {
      result.push([key, value]);
    }
  }
  return result;
};
var toWebRequest = (nodeReq, urlPath) => {
  const { method, headers, url = "" } = nodeReq;
  const ac = new AbortController();
  const origin = `http://${headers.host}`;
  const fullUrl = new URL(urlPath ?? url, origin);
  const webStream = Readable.toWeb(nodeReq);
  nodeReq.once("aborted", () => ac.abort());
  return new Request(fullUrl.href, {
    method,
    headers: normalizeHeaders(headers),
    body: method === "GET" || method === "HEAD" ? null : webStream,
    // @ts-expect-error Not typed
    duplex: "half"
  });
};
var fromWebResponse = async (webRes, res) => {
  res.statusCode = webRes.status;
  webRes.headers.forEach((value, name) => {
    res.setHeader(name, value);
  });
  if (webRes.body) {
    const reader = webRes.body.getReader();
    const writer = res;
    while (true) {
      const { done, value } = await reader.read();
      if (done) {
        break;
      }
      writer.write(value);
    }
  }
  res.end();
};

// src/main.ts
function netlify(options = {}) {
  if (process.env.NETLIFY_DEV) {
    return [];
  }
  const plugin = {
    name: "vite-plugin-netlify",
    async configureServer(viteDevServer) {
      const { blobs, functions, middleware = true, redirects, staticFiles } = options;
      const netlifyDev = new NetlifyDev({
        blobs,
        functions,
        logger,
        redirects,
        staticFiles,
        projectRoot: viteDevServer.config.root
      });
      await netlifyDev.start();
      if (!netlifyDev.siteIsLinked) {
        logger.log("Your project is not linked to a Netlify site. Run `npx netlify link` to get started.");
      }
      if (middleware) {
        return () => {
          viteDevServer.middlewares.use(async (nodeReq, nodeRes, next) => {
            const req = toWebRequest(nodeReq, nodeReq.originalUrl);
            const res = await netlifyDev.handle(req);
            if (res) {
              fromWebResponse(res, nodeRes);
            } else {
              next();
            }
          });
        };
      }
    }
  };
  return [plugin];
}
export {
  netlify as default
};
